research/spoofax-org-export.tex

```diff
@@ -0,0 +1,234 @@
+specification of, for example, Standard ML~\cite{Milner97}. The
+lexical grammar is most often defined using regular expression. It
+defines the individual words made up of characters, such as
+identifiers and numeric constants. The context-free grammar then
+defines syntactically valid sentences made up of words.
+
+\subsubsection{SDF3: syntax definition in Spoofax}
+\label{sec:orgheadline1}
+To specify a syntax definition declaratively in Spoofax, a DSL called
+\emph{SDF3}~\cite{Vollebregt12} is used, the third generation of the
+\emph{Syntax Definition Formalism} (SDF)~\cite{Heering89}. SDF3 uses
+context-free grammer productions for the specification of both the
+lexical syntax and the context-free syntax, a feature that was
+introduced in SDF2~\cite{Visser97}.
+
+The declarative nature of SDF3 allows for thinking in terms of the
+structure (the \emph{what}), instead of in terms of parser algorithms (the
+\emph{how}) as is the case with many current parsing
+algorithms~\cite{Kats10b}. The syntax definition can be used to
+make parsers which parse a textual representation to an AST, and
+pretty-printers for mapping ASTs back to text. However, due to its
+declarative nature, SDF3 is not limited to generating parsers and
+pretty printers: it can also be used for error recovery
+rules~\cite{deJonge12}, syntax highlighting rules and folding
+rules for editors.
+
+All of the other parts of Spoofax use the AST that is produced by a
+parser generated from a SDF3 specification.
+\subsection{Static Semantics}
+\label{sec-static-analysis}
+Static semantics refer to the meaning of what is a well-formed program
+for a particular language~\cite{Milner97}. This imposes more
+constraints than syntax definition, such as name binding, scoping
+rules, and type checking. These cannot be specified by a syntax
+definition alone, and are thus considered separately.
+\subsubsection{Declarative static semantics specification in Spoofax}
+\label{sec:orgheadline2}
+In Spoofax, all the static semantics as well as the dynamic semantics
+used to be specified with the \emph{Stratego} transformation language
+(which is discussed in section \ref{sec-term-rewrite}). Nowadays, two
+high-level DSLs exist for specifying static semantics declaratively:
+NaBL and TS. The two DSLs can work together: for instance, the type of
+a variable can be set with NaBL, so that TS can be used to make
+assertions on the type of that variable.
+\subsubsection{NaBL: the Name Binding Language}
+\label{sec:orgheadline3}
+With \emph{NaBL} (pronounced \emph{enable}), name binding and scoping rules can
+be specified declaratively~\cite{KonatKWV12}. Here is an example
+of the name binding and scoping rules for a class, from the \emph{paplj}
+language\footnote{paplj is used as an exercise language for the
+"Declare Your Language" book, which is a work-in-progress at the time
+of writing. More information can be found here:
+\url{https://github.com/MetaBorgCube/declare-your-language}}
+\begin{verbatim}
```

Comment:
Perhaps it's worth the time to write a new syntax highlighting thingy for the lstlistings package. (It can be used again in the final report).